File: services/event-sourcing/event_store.ts
export interface Event {
  id: string;
  aggregateId: string;
  aggregateType: string;
  eventType: string;
  data: any;
  metadata: Record<string, any>;
  version: number;
  timestamp: Date;
}

export interface Snapshot {
  aggregateId: string;
  aggregateType: string;
  state: any;
  version: number;
  timestamp: Date;
}

export class EventStore {
  private events: Event[] = [];
  private snapshots: Map<string, Snapshot> = new Map();
  private snapshotInterval = 10;

  async appendEvent(event: Omit<Event, 'id' | 'timestamp'>): Promise<Event> {
    const newEvent: Event = {
      ...event,
      id: this.generateId(),
      timestamp: new Date(),
    };

    this.events.push(newEvent);

    if (newEvent.version % this.snapshotInterval === 0) {
      await this.createSnapshot(newEvent.aggregateId, newEvent.aggregateType);
    }

    return newEvent;
  }

  async getEvents(
    aggregateId: string,
    fromVersion: number = 0
  ): Promise<Event[]> {
    return this.events.filter(
      (event) =>
        event.aggregateId === aggregateId && event.version > fromVersion
    );
  }

  async getEventsByType(
    aggregateType: string,
    eventType?: string
  ): Promise<Event[]> {
    return this.events.filter(
      (event) =>
        event.aggregateType === aggregateType &&
        (!eventType || event.eventType === eventType)
    );
  }

  async getSnapshot(aggregateId: string): Promise<Snapshot | null> {
    return this.snapshots.get(aggregateId) || null;
  }

  async createSnapshot(aggregateId: string, aggregateType: string): Promise<void> {
    const events = await this.getEvents(aggregateId);

    if (events.length === 0) return;

    const state = this.replayEvents(events);
    const lastEvent = events[events.length - 1];

    const snapshot: Snapshot = {
      aggregateId,
      aggregateType,
      state,
      version: lastEvent.version,
      timestamp: new Date(),
    };

    this.snapshots.set(aggregateId, snapshot);
  }

  async loadAggregate(aggregateId: string): Promise<any> {
    const snapshot = await this.getSnapshot(aggregateId);

    const fromVersion = snapshot ? snapshot.version : 0;
    const events = await this.getEvents(aggregateId, fromVersion);

    let state = snapshot ? snapshot.state : {};

    if (events.length > 0) {
      state = this.replayEvents(events, state);
    }

    return state;
  }

  private replayEvents(events: Event[], initialState: any = {}): any {
    return events.reduce((state, event) => {
      return this.applyEvent(state, event);
    }, initialState);
  }

  private applyEvent(state: any, event: Event): any {
    switch (event.eventType) {
      case 'UserCreated':
        return {
          ...state,
          id: event.aggregateId,
          ...event.data,
        };

      case 'UserUpdated':
        return {
          ...state,
          ...event.data,
        };

      case 'UserDeleted':
        return {
          ...state,
          deleted: true,
        };

      default:
        return state;
    }
  }

  private generateId(): string {
    return `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
  }

  async getEventStream(
    aggregateType?: string,
    fromTimestamp?: Date
  ): Promise<Event[]> {
    let filtered = this.events;

    if (aggregateType) {
      filtered = filtered.filter((e) => e.aggregateType === aggregateType);
    }

    if (fromTimestamp) {
      filtered = filtered.filter((e) => e.timestamp >= fromTimestamp);
    }

    return filtered;
  }
}

export class EventBus {
  private handlers: Map<string, Array<(event: Event) => void>> = new Map();

  subscribe(eventType: string, handler: (event: Event) => void): void {
    if (!this.handlers.has(eventType)) {
      this.handlers.set(eventType, []);
    }

    this.handlers.get(eventType)!.push(handler);
  }

  async publish(event: Event): Promise<void> {
    const handlers = this.handlers.get(event.eventType) || [];

    for (const handler of handlers) {
      await handler(event);
    }
  }
}

File: services/ruby/event_sourcing/event_store.rb
module EventSourcing
  class Event
    attr_reader :aggregate_id, :type, :data, :metadata, :timestamp, :version

    def initialize(aggregate_id:, type:, data:, metadata: {}, version: 1)
      @aggregate_id = aggregate_id
      @type = type
      @data = data
      @metadata = metadata
      @timestamp = Time.now
      @version = version
    end

    def to_h
      {
        aggregate_id: @aggregate_id,
        type: @type,
        data: @data,
        metadata: @metadata,
        timestamp: @timestamp,
        version: @version
      }
    end
  end

  class EventStore
    def initialize
      @events = []
      @snapshots = {}
    end

    def append(event)
      @events << event
      publish_event(event)
    end

    def get_events(aggregate_id, from_version: 0)
      @events.select do |event|
        event.aggregate_id == aggregate_id && event.version > from_version
      end
    end

    def get_all_events(from: 0, to: nil)
      events = @events[from..-1] || []
      to ? events[0...to] : events
    end

    def create_snapshot(aggregate_id, state, version)
      @snapshots[aggregate_id] = {
        state: state,
        version: version,
        timestamp: Time.now
      }
    end

    def get_snapshot(aggregate_id)
      @snapshots[aggregate_id]
    end

    def replay(aggregate_id)
      snapshot = get_snapshot(aggregate_id)
      
      if snapshot
        state = snapshot[:state]
        from_version = snapshot[:version]
      else
        state = {}
        from_version = 0
      end

      events = get_events(aggregate_id, from_version: from_version)
      
      events.each do |event|
        state = apply_event(state, event)
      end

      state
    end

    private

    def apply_event(state, event)
      case event.type
      when 'user_created'
        state.merge(event.data)
      when 'user_updated'
        state.merge(event.data)
      when 'post_created'
        state[:posts] ||= []
        state[:posts] << event.data
        state
      else
        state
      end
    end

    def publish_event(event)
      # Publish to event bus for subscribers
      puts "Event published: #{event.type} for #{event.aggregate_id}"
    end
  end

  class AggregateRoot
    attr_reader :id, :version, :uncommitted_events

    def initialize(id)
      @id = id
      @version = 0
      @uncommitted_events = []
    end

    def apply_event(event)
      @uncommitted_events << event
      @version += 1
    end

    def commit_events(event_store)
      @uncommitted_events.each do |event|
        event_store.append(event)
      end
      @uncommitted_events.clear
    end

    def load_from_history(events)
      events.each do |event|
        apply(event)
        @version = event.version
      end
    end

    protected

    def apply(event)
      # Override in subclasses
    end
  end
end
